classes/core/PKPRequest.php

```diff
@@ -103,6 +103,13 @@ public function setDispatcher($dispatcher)
      */
     public function &getDispatcher()
     {
+        if ( ! $this->_dispatcher ) {
```

Comment:
This potentially an invasive change. Instead, I think we should stop depending on `PKPRequest` in areas of the code that may be called e.g. via command line. A common problem is URL generation; the usual pattern, `$request->url`, depends on rather than calling `$request->url`, for example, you can use instead:
```php
$request = $application->getRequest();
$dispatcher = $application->getDispatcher();
$url = $dispatcher->url($request, PKPApplication::ROUTE_PAGE, ...);
```
This still depends on `$request` being fetched and passed in -- we have more work to do around disentangling this -- but it'll work when called in a command-line environment. This is done e.g. in `lib/pkp/classes/task/ReviewReminder.php`.

Reply:
I agree 100% and this is not an ideal implementation but more like a patch up or hot fix . But the problem lies with our excessive use of code such as 
```php
Application::get()->getRequest()->getDispatcher()
```

if we have have even used it like 
```php
Application::get()->getDispatcher()
```

it would have been ok or the way you have mentioned . Now if this does not get included , then we have to refactor all the occurrence of such code blocks to make it compatible with queue jobs or any command line utility that may run on separate process outside of main app process . I can see one potential place for such refactoring is in the `mailable` section . Also others also need to be careful about code implementation so that it would not break if that implementation get associated with a queue jobs through worker. I feel like most of us continue to use approach like `Application::get()->getRequest()->getDispatcher()` and get stuck with queue jobs to figure out why it's not working or throwing internal exception . Also our job ui is not sophisticated yet to debug failed jobs from there( that is the next part I want to work on related to queue jobs) . 

One example of such case already is job `PKP\Jobs\Email\EditorialReminder` failed when running it through the job command line tool as the core `mailable` and `email variable classes` have call such as `Application::get()->getRequest()->getDispatcher()`. 

@asmecher  In fact I did not like that implementation of mine either and have nothing against removing that but doing so  we will be forcing others to do a lot of refactoring probably. what do you think about it ? if we want to avoid such refactoring now , how about keep it for now, let others know about such cases, after a certain time we remove it when we are sure that such calls have been removed from our codebase , what you say ?

Reply:
Thinking a little more about this -- I see your point. And the more verbose work-around I suggested above...
```php
$request = $application->getRequest();
$dispatcher = $application->getDispatcher();
$url = $dispatcher->url($request, PKPApplication::ROUTE_PAGE, ...);
```
...is hardly less dependent on the `$request` object. So let's go ahead with your proposal!